### DISCUSSION TITLE
동적타입언어인 자바스크립트의 특징을 말해주세요.

### DISCUSSION THREAD

**Discussion by pipisebastian** [publdaze/js-study#9] [데이터 타입]
- 동적타입언어와 정적타입언어를 비교해서 설명해주세요.
  - JavaScript, TypeScript을 비교해서 설명해주세요.
  - 프로젝트에서 어떤 언어를 사용해야한다고 생각하시나요?

**Comment by publdaze** [comment]
> - 동적타입언어와 정적타입언어를 비교해서 설명해주세요.
- 동적 타입 언어는 런타임에 타입 결정
- 정적 타입 언어는 컴파일 타임에 타입 결정 -> 변수 선언 시 타입 미리 선언 필요
>  - JavaScript, TypeScript을 비교해서 설명해주세요.
  - javaScript는 동적 타입 언어
  - typeScript는 정적 타입 언어
>  - 프로젝트에서 어떤 언어를 사용해야한다고 생각하시나요?
  - 자유도 높게 사용하려면 js, 안정성 높게 사용하려면 ts
  - 나는 ts 선호하는 편
    - 프로젝트 경험 - 리뉴얼1 유지 보수하다가 객체 속성이 존재하지 않은 상태에 접근하여 발생하는 버그 다수 존재 -> 리뉴얼2에 ts 도입하여 안정성 확보

**Comment by pipisebastian** [comment]
>동적타입언어와 정적타입언어를 비교해서 설명해주세요. JavaScript, TypeScript을 비교해서 설명해주세요.

1. 정적타입언어 (c, c++, java, TypeScript)
    1. 컴파일 시점에서 타입 체크. 불일치할 경우 실행자체를 막음
        1. 컴파일 :  문법 에러, 타입 체크
    2. 에러 : `컴파일`, `유연성 감소`, `신뢰도 증가`, 
2. 동적타입언어 (JavaScript)
    1. 런타임 시점, 소스코드를 실행하며, 값을 **할당**하는 시점에서 **타입이 결정**. 재할당에 의해 타입이 바뀜
        1. 런타임 : 메모리 부족
    2. 에러 : `런타임`, `유연성 증가`, `신뢰도 감소`  
    

>프로젝트에서 어떤 언어를 사용해야한다고 생각하시나요?

프로젝트 규모와 사용하는 언어에 따라 다르게 선택할 것 같습니다.
프로젝트 규모가 작거나, 이미 **JavaScript로** 개발이 많이 진행되어있다면, 속도와 유연성을 위해 자바스크립트를 선택할 것 같습니다.
그러나, 프로젝트 규모가 크고, 안정성이 중요하다면 TypeScript를 사용할 것 같습니다.

개인적으로 코드를 작성할 때 IDE에서 자동 완성 기능을 제공하고, 타입 관련 에러를 사전에 파악할 수 있기에 타입스크립트를 선호하는 편입니다.
